network/wsPeer.go

```diff
@@ -525,3 +602,75 @@ func (wp *wsPeer) CheckSlowWritingPeer(now time.Time) bool {
 	timeSinceMessageCreated := now.Sub(time.Unix(0, ongoingMessageTime))
 	return timeSinceMessageCreated > maxMessageQueueDuration
 }
+
+// getNonce returns the byte representation of ever increasing uint64
+// The value is stored on wsPeer
+func (wp *wsPeer) getNonce() []byte {
+	atomic.AddUint64(&wp.nonce, 1)
+	buf := make([]byte, binary.MaxVarintLen64)
+	binary.PutUvarint(buf, wp.nonce)
+	return buf
+}
+
+// Request submits the request to the server, waits for a response
+func (wp *wsPeer) Request(ctx context.Context, tag Tag, topics Topics) (resp *Response, e error) {
+
+	// Add nonce as a topic
+	nonce := wp.getNonce()
+	topics = append(topics, Topic{key: "nonce", data: nonce})
+
+	// serialize the topics
+	serializedMsg := topics.MarshallTopics()
+
+	// Get the topics' hash
+	hash := Hash(serializedMsg)
+
+	// Make a response channel to wait on the server response
+	wp.makeResponseChannel(hash)
+	defer wp.removeResponseChannel(hash)
+
+	// Send serializedMsg
+	select {
+	case wp.sendBufferBulk <- sendMessage{
+		data:         append([]byte(tag), serializedMsg...),
+		enqueued:     time.Now(),
+		peerEnqueued: time.Now()}:
+	case <-wp.closing:
+		wp.net.log.Debugf("peer closing %s", wp.conn.RemoteAddr().String())
+		return
+	case <-ctx.Done():
+		return resp, ctx.Err()
+	}
+
+	// wait for the channel.
+	select {
+	case resp = <-wp.responseChannels[hash]:
+		return resp, nil
+	case <-wp.closing:
+		wp.net.log.Debugf("peer closing %s", wp.conn.RemoteAddr().String())
+		return
```

Comment:
you need to exit here with an error so that the caller won't attempt to use the `resp`. In go, when there is a `something, error` returned from a function, there is expectation that if the latter is nil then the former is valid.